[clangd] Prevent printing huge initializer lists in hover definitions

rwols opened the pull request:
Previously we checked whether we were dealing with a large initializer using TokenBuffer::expandedTokens. However, TokenBuffer does not contain the tokens of the preamble. This causes large arrays imported from an #include to still print all their elements. The added unit test checks the case of a large array included from the preamble.

kadircet commented on clang-tools-extra/clangd/Hover.cpp:
```diff
@@ -138,15 +138,9 @@ std::string getNamespaceScope(const Decl *D) {
-  }
+  // Initializers might be huge and result in lots of memory allocations in some
+  // catostrophic cases. Such long lists are not useful in hover cards anyway.
+  PP.EntireContentsOfLargeArray = false;
```
i am not sure if this completely addresses the previous implementation. e.g. we can have a nested initializer list, which never has more than N elements directly, but the overall initalizer list might be huge. in such a scenario, we'll still end up allocating lots of strings.

so can we keep the old filtering while also setting this flag here? note that it'll still be ~incomplete. a more concrete approach could be based on counting number of "total" elements in the initializer list, and setting the suppression flag based on that. i'd actually lean towards such a solution, WDYT?

rwols replied:
> a more concrete approach could be based on counting number of "total" elements in the initializer list, and setting the suppression flag based on that. i'd actually lean towards such a solution, WDYT?

I agree, but don't know where to start on this. Since TokenBuffer doesn't have tokens from the preamble, what alternatives are there?

kadircet replied:
I wasn't suggesting to use tokens, but rather count the number of elements directly through the AST.
we've got the expression available already, so we can go over its `chidlren()` and count total elements we **might** encounter, and suppress intializers based on that (when the tokens aren't available)

rwols replied:
OK, I check the size of the children() now when the initializer expression is not in the main file, and added two unit tests that verify that for a large initializer list from a header file it won't render, while a "small" initializer list will still render.

kadircet commented on clang/lib/AST/StmtPrinter.cpp:
```diff
@@ -1720,6 +1720,12 @@ void StmtPrinter::VisitInitListExpr(InitListExpr* Node) {
     if (i) OS << ", ";
+    // TODO: There is duplicated functionality in APValue::printPretty.
+    // Would be good to consolidate the two.
+    if (!Policy.EntireContentsOfLargeArray && i == 10) {
```
i am not sure if usage of this flag is actually appropriate here. it specifically says "contents of large arrays" but here we have an initializer list, not necessarily an array.

moreover this fix is in clang, independent of clangd, so I think it deserves its own patch, with relevant owners from clang. can you separate it out? i'd like to hear their opinion on using this flag.

rwols replied:
I reverted these changes in the clang namespace.

rwols committed:
[clangd] Do not render large initializer expressions from the preamble

An attempt is made to estimate the size of the initializer expression.
If it has less than 100 (in)direct AST child nodes we'll assume it should
render OK.
